apps/web/src/components/AssessmentStepTracker/AssessmentStepTracker.tsx

```diff
@@ -22,34 +18,88 @@ import {
   groupPoolCandidatesByStep,
   defaultFilters,
   filterAlreadyDisqualified,
+  generateStepName,
 } from "./utils";
 import Filters from "./Filters";
+import SpinnerIcon from "../SpinnerIcon/SpinnerIcon";
 
 const talentPlacementLink = (chunks: React.ReactNode, href: string) => (
   <Link href={href}>{chunks}</Link>
 );
-export interface AssessmentStepTrackerProps {
-  pool: Pool;
-}
 
-const generateStepName = (step: AssessmentStep, intl: IntlShape): string => {
-  // check if title exists in LocalizedString object, then return empty string if not for a truthy check
-  const titleLocalized = getLocalizedName(step.title, intl, true);
-  if (titleLocalized) {
-    return titleLocalized;
+export const AssessmentStepTracker_CandidateFragment = graphql(/* GraphQL */ `
+  fragment AssessmentStepTracker_Candidate on PoolCandidate {
+    id
+    isBookmarked
+    status
+    pool {
+      id
+    }
+    user {
+      id
+      firstName
+      lastName
+      armedForcesStatus
+      hasPriorityEntitlement
+    }
+    assessmentResults {
+      id
+      assessmentStep {
+        id
+      }
+      assessmentDecision
+      assessmentDecisionLevel
+      assessmentResultType
+      poolSkill {
+        id
+        type
+      }
+    }
   }
-  if (step.type) {
-    return intl.formatMessage(getAssessmentStepType(step.type));
+`);
+
+export const AssessmentStepTracker_PoolFragment = graphql(/* GraphQL */ `
+  fragment AssessmentStepTracker_Pool on Pool {
+    id
+    assessmentSteps {
+      id
+      title {
+        en
+        fr
+      }
+      type
+      sortOrder
+      poolSkills {
+        id
+        type
+      }
+    }
   }
-  return intl.formatMessage(commonMessages.notAvailable);
-};
+`);
 
-const AssessmentStepTracker = ({ pool }: AssessmentStepTrackerProps) => {
+export interface AssessmentStepTrackerProps {
+  poolQuery?: FragmentType<typeof AssessmentStepTracker_PoolFragment>;
+  candidateQuery?: FragmentType<
+    typeof AssessmentStepTracker_CandidateFragment
+  >[];
+  fetching: boolean;
+}
+
+const AssessmentStepTracker = ({
+  poolQuery,
+  candidateQuery,
+  fetching,
+}: AssessmentStepTrackerProps) => {
   const intl = useIntl();
   const paths = useRoutes();
   const [filters, setFilters] = React.useState<ResultFilters>(defaultFilters);
-  const steps = unpackMaybes(pool.assessmentSteps);
-  const candidates = unpackMaybes(pool.poolCandidates);
+  const pool = getFragment(AssessmentStepTracker_PoolFragment, poolQuery);
+  const steps = unpackMaybes(pool?.assessmentSteps);
+  const poolCandidates = getFragment(
+    AssessmentStepTracker_CandidateFragment,
+    candidateQuery,
+  );
+  const candidates = unpackMaybes(poolCandidates ? [...poolCandidates] : []);
   // to handle partially completed processes, filter out candidates assessed and disqualified pre-RoD
   const filteredCandidates = filterAlreadyDisqualified(candidates);
   const groupedSteps = groupPoolCandidatesByStep(steps, filteredCandidates);
```

Comment:
I'd like to adjust the typing of some of these functions to be more definite to go along with your fragmenting addition. As this is some complicated stuff acting on quite a few fields, it is hard to catch a missed field in queries. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/de5d7f50-c6e7-4e98-96cf-263d10d4706e)

Anything complicated but worth it can be spun off. 
I was thinking for example, the `filterAlreadyDisqualified` function as seen above could be narrowed to be type safe.

Instead of the whole PoolCandidate object, it just acts on the status field and that candidates had their assessment results fetched. Type safe would then like 
```
{
status: PoolCandidateStatus; 
assessmentResults: AssessmentResult [];
} []
```
I think. So a place with the lite version of typing we liked. 

![image](https://github.com/GCTC-NTGC/gc-digital-talent/assets/40485260/894b5261-791b-4c47-8f3f-7702922c8dd3)

Reply:
So, while this is a good suggestion. We discussed and I believe a lot of this code will be shifted to the server so I'm not sure how worth adjusting the types are right now. I think we plan on creating a query where you can do:

```graphql
query {
  pool(id: $id) {
     assessmentSteps {
       poolCandidates {
          ...AssessmentStepTrackerCandidate
       }
     }
  }
}
```

At the same time, I believe we will be moving a lot of the business logic (`filterAlreadyDisqualified`, `groupPoolCandidatesByStep`) onto the server instead. That way we only get the candidates that "matter".

Reply:
Sure, if you think it will be unneeded, fine by me